protocol/auditor.go

```diff
@@ -0,0 +1,176 @@
+// or an auditor's pinned signing key in its history.
+func (a *AudState) verifySTRConsistency(prevSTR, str *DirSTR) error {
+	// verify STR's signature
+	if !a.signKey.Verify(str.Serialize(), str.Signature) {
+		return CheckBadSignature
+	}
+	if str.VerifyHashChain(prevSTR) {
+		return nil
+	}
+
+	// TODO: verify the directory's policies as well. See #115
+	return CheckBadSTR
+}
+
+// checkSTRAgainstVerified checks an STR str against the a.verifiedSTR.
+// If str's Epoch is the same as the verified, checkSTRAgainstVerified()
+// compares the two STRs directly. If str is one epoch ahead of the
+// a.verifiedSTR, checkSTRAgainstVerified() checks the consistency between
+// the two STRs.
+// checkSTRAgainstVerified() returns nil if the check passes,
+// or the appropriate consistency check error if any of the checks fail,
+// or str's epoch is anything other than the same or one ahead of
+// a.verifiedSTR.
+func (a *AudState) checkSTRAgainstVerified(str *DirSTR) error {
+	// FIXME: check whether the STR was issued on time and whatnot.
+	// Maybe it has something to do w/ #81 and client
+	// transitioning between epochs.
+	// Try to verify w/ what's been saved
+
+	// FIXME: we are returning the error immediately
+	// without saving the inconsistent STR
+	// see: https://github.com/coniks-sys/coniks-go/pull/74#commitcomment-19804686
+	switch {
+	case str.Epoch == a.verifiedSTR.Epoch:
+		// Checking an STR in the same epoch
+		if err := a.compareWithVerified(str); err != nil {
+			return err
+		}
+	case str.Epoch == a.verifiedSTR.Epoch+1:
+		// Otherwise, expect that we've entered a new epoch
+		if err := a.verifySTRConsistency(a.verifiedSTR, str); err != nil {
+			return err
+		}
+	default:
+		return CheckBadSTR
+	}
+
+	return nil
+}
+
+// verifySTRRange checks the consistency of a range
+// of a directory's STRs. It begins by verifying the STR consistency between
+// the given prevSTR and the first STR in the given range, and
+// then verifies the consistency between each subsequent STR pair.
+func (a *AudState) verifySTRRange(prevSTR *DirSTR, strs []*DirSTR) error {
+	prev := prevSTR
+	for i := 0; i < len(strs); i++ {
+		str := strs[i]
+		if str == nil {
+			// FIXME: if this comes from the auditor, this
+			// should really be an ErrMalformedAuditorMessage
+			return ErrMalformedDirectoryMessage
+		}
+
+		// verify the consistency of each STR in the range
+		if err := a.verifySTRConsistency(prev, str); err != nil {
+			return err
+		}
+
+		prev = str
+	}
+
+	return nil
+}
+
+// AuditDirectory validates a range of STRs received from a CONIKS directory.
+// AuditDirectory() checks the consistency of the oldest STR in the range
+// against the verifiedSTR, and verifies the remaining
+// range if the message contains more than one STR.
+// AuditDirectory() returns the appropriate consistency check error
+// if any of the checks fail, or nil if the checks pass.
+func (a *AudState) AuditDirectory(strs []*DirSTR) error {
```

Comment:
Hmm, I generally assume that the caller hasn't validated the parameters. In the case of `AuditDirectory`, both `ConsistencyChecks` and `AuditLog`, both will call it inside of `Audit`, so in that case I think it would be safe to assume that the caller did a validation beforehand. OTOH, adding a parameter validation here is trivial, so I think we can add it here to be safe.